travertino/tests/style/test_composite.py

```diff
@@ -0,0 +1,167 @@
+import pytest
+
+from .style_classes import VALUE1, VALUE2, VALUE3, VALUE4, Style
+
+
+def test_default_values():
+    """Shorthand property returns the initial defaults."""
+    style = Style()
+    assert style.composite_no_optional == (VALUE1, 0)
+    assert style.composite_optional == (None, VALUE2, VALUE1, 0)
+
+
+@pytest.mark.parametrize(
+    "prop_name, min_num, max_num",
+    [
+        ("composite_no_optional", 2, 2),
+        ("composite_optional", 2, 4),
+    ],
+)
+def test_wrong_number_args(prop_name, min_num, max_num):
+    """Error when too many or not enough arguments."""
+    for num_values in [min_num - 1, max_num + 1]:
+        values = (VALUE1,) * num_values
+
+        with pytest.raises(
+            TypeError,
+            match=(
+                rf"Composite property '{prop_name}' must be set with at least "
+                rf"{min_num} and no more than {max_num} values\."
+            ),
+        ):
+            _ = Style(**{prop_name: values})
+
+
+def test_set_composite_with_no_optionals():
+    """Values can be set and retrieved on a shorthand with only required properties."""
+    style = Style(composite_no_optional=(VALUE2, VALUE3))
+    assert style.composite_no_optional == (VALUE2, VALUE3)
+    assert style.explicit_const == VALUE2
+    assert style.explicit_value == VALUE3
+
+
+# The following all test the logic involved with optional properties.
+
+
+def assert_composite(style, expected):
+    # Test against retrieving the composite property
+    assert style.composite_optional == expected
+
+    # Also test against the referenced properties directly.
+    (explicit_none, different_values_prop, explicit_const, explicit_value) = expected
+    assert style.explicit_none == explicit_none
+    assert style.different_values_prop == different_values_prop
+    assert style.explicit_const == explicit_const
+    assert style.explicit_value == explicit_value
+
+
+def test_assign_only_requied():
+    """Assigning only the required properties works."""
+    style = Style(composite_optional=(VALUE2, VALUE3))
+
+    assert_composite(style, (None, VALUE2, VALUE2, VALUE3))
+    assert "different_values_prop" not in style
+    assert "explicit_none" not in style
```

Comment:
This test looks completely correct, but could do with some explanatory notes so it's clear what it's testing.